### PR TITLE
Fix and improve Watcher.detectWatcher

### PR DESCRIPTION
**Pull request by stefanpenner**
- [x] add tests to detectWatcher
- [x] improve fallback scenarios (not all systems node works on support fs.watch)
- [x]  make sure we consistently do the right thing, regardless of which branches we take.
- [x] windows CI seems to be failing for unknown reasons....

Ok, so this refactors the watcher detection code (which was quite complicated, untestable, buggy, and nearly impossible to extend without causing more issues). The goal here is, a separate model which can be tested, extended and improved with confidence. 

I feel This has festered long enough, a refactor was long overdue.

This PR also adds the ability to detect if `node` watcher extremely broken (we could extend this to detect if change events are even possible, but that can come later). If it is extremely broken, we log it and fall back to `polling`. This hopefully will both provide useful information to the user, and attempt a grace-full fallback.

### CODE REVIEW

File: lib/models/watch-detector.js [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  };
+}
+
+WatchPreference.prototype.watchmanWorks = function(details) {
+  this._watchmanInfo.enabled = true;
+  this._watchmanInfo.version = details.version;
+  this._watchmanInfo.canNestRoots = semver.satisfies(details.version, '>= 3.7.0');
+};
+
+Object.defineProperty(WatchPreference.prototype, 'watchmanInfo', {
```
Comment (rwjblue) [inline lib/models/watch-detector.js]: What do we gain by adding this as a getter?  Just the ability to change in the future?

File: lib/models/watch-detector.js [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+  get: function() {
+    return this._watchmanInfo;
+  }
+});
+
+module.exports = WatchDetector;
```
Comment (rwjblue) [inline lib/models/watch-detector.js]: This always breaks by brain.  When an export is above the function.  Yes, I know hoisting is a thing, but still my brain hurts.

File: lib/models/watch-detector.js [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    return new WatchPreference(WATCHMAN);
+  }
+}
+
+/*
+ * @public
+ *
+ * Detect if `fs.watch` provided by node is capable of observing a directory
+ * within `process.cwd()`. Although `fs.watch` is provided as a part of the
+ * node stdlib, their exists platforms it ships to that do not implemented the
```
Comment (rwjblue) [inline lib/models/watch-detector.js]: s/implemented/implement/

File: lib/models/watch-detector.js (hunk 1 above)
Comment (stefanpenner) [inline lib/models/watch-detector.js]: can't set to it from the outside. (its readOnly)

File: lib/models/watch-detector.js [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+ * required sub-system. For example, there exists `posix` targets, without both
+ * inotify and FSEvents
+ *
+ * @method testIfNodeWatcherAppearsToWork
+ * @returns {Boolean) outcome
+*/
+WatchDetector.prototype.testIfNodeWatcherAppearsToWork  = function() {
+  return new Promise(function(resolve) {
+    var key = path.dirname(path.dirname(process.cwd()));
```
Comment (rwjblue) [inline lib/models/watch-detector.js]: Is there a specific reason to go two directories up? This is somewhat confusing to me.  I would think that you would want to use the actual `project.root` which is generally `process.cwd()`.

File: lib/models/watch-detector.js [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+      return;
+    }
+
+    this._doesNodeWorkCache[key] = false;
+    try {
+      // builds a tmp directory at process.cwd() + '/tmp/' + 'something-unique';
+      var tmpDir = quickTemp.makeOrRemake(this, 'tmp');
+
+      this.fs.watch(tmpDir, { persistent: false, recursive: false });
+      this.fs.unwatch(tmpDir);
```
Comment (rwjblue) [inline lib/models/watch-detector.js]: Should we actually change the file, and confirm it works?

File: lib/models/watch-detector.js [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+ *  * watchman executable may be something unexpected
+ *  * ???
+ *
+ *  @method checkWatchman
+ *  @returns {Object} watch preference + _watchmanInfo
+*/
+WatchDetector.prototype.checkWatchman = function() {
+  var result = new WatchPreference(null);
+
+  return this.childProcess.exec('watchman version').then(function(output) {
```
Comment (rwjblue) [inline lib/models/watch-detector.js]: Might want a `debug` for this check, and its result (including the error case when non-JSON is emitted)

File: tests/unit/models/watch-detector-test.js [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        expect(ui.output).to.not.match(/Could not start watchman/);
+        expect(ui.output).to.not.match(/falling back to NodeWatcher/);
+        expect(ui.output).to.not.match(/ember-cli\.com\/user-guide\/\#watchman/);
+        expect(ui.output).to.not.match(/Looks like you have a different program called watchman/);
+        expect(ui.output).to.not.match(/Invalid watchman found/);
+      });
+    });
+
+    describe('fallse back to NODE', function() {
+      var iff = it;
```
Comment (rwjblue) [inline tests/unit/models/watch-detector-test.js]: O_o

**Comment by rwjblue** [comment]
This looks great! I left a few comments, but mostly just things I was thinking as I reviewed.  Only one that seems like it may be wrong (though you may correct me here).

File: tests/unit/models/watch-detector-test.js (hunk 7 above)
Comment (stefanpenner) [inline tests/unit/models/watch-detector-test.js]: the lines bellow this read better as "iff(xyz)"... "if and only if xyz"

File: lib/models/watch-detector.js (hunk 4 above)
Comment (stefanpenner) [inline lib/models/watch-detector.js]: Ya, going up 2 gives us the stable directory in all cases, But should be explore further in app usage (rather then test usage scenarios)

File: tests/unit/models/watch-detector-test.js (hunk 7 above)
Comment (rwjblue) [inline tests/unit/models/watch-detector-test.js]: heh, they only read better if you know that `iif` means "if and only if". You CS people. 😈

**Comment by stefanpenner** [comment]
> This looks great! I left a few comments, but mostly just things I was thinking as I reviewed. Only one that seems like it may be wrong (though you may correct me here).

https://en.wikipedia.org/wiki/If_and_only_if apparently its MATHs people

**Comment by rwjblue** [comment]
HAHA, MATHS! 🍇

File: lib/models/watch-detector.js (hunk 5 above)
Comment (stefanpenner) [inline lib/models/watch-detector.js]: some future PR could, this one merely makes sure the call does throw. Like it does on platforms that don't have support.

File: lib/models/watch-detector.js (hunk 4 above)
Comment (stefanpenner) [inline lib/models/watch-detector.js]: I'm going to re-work this yet.

**Comment by stefanpenner** [comment]
the following appears to be possible, but should not.
```
was unable to use: "events", fell back to: "watchman"
```

- [ ] fix this

**Comment by stefanpenner** [comment]
crap forgot about this, must ship soon.

File: lib/models/watch-detector.js (hunk 4 above)
Comment (stefanpenner) [inline lib/models/watch-detector.js]: update to be project.root

**Comment by stefanpenner** [comment]
lets see what CI says.

**Comment by jonnii** [comment]
https://www.youtube.com/watch?v=AJQ3TM-p2QI

**Comment by stefanpenner** [comment]
Looks like one of the windows CI runs passed, but one failed (node 4.x) but failed due to connection \w github issue. 

Re-running...

**Comment by stefanpenner** [comment]
Looks like all is green. Any reviewers ?

**Pull request merged**